azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/AbstractAzureMojo.java

```diff
@@ -272,74 +274,97 @@ public int getHttpProxyPort() {
+                .filter(t -> StringUtils.equalsIgnoreCase(t.getSubscriptionId(), defaultId)).findFirst().orElse(null);
+
+        final SubscriptionOption subscriptionOptionSelected = new CustomTextIoStringListReader<SubscriptionOption>(() -> textIO.getTextTerminal(), null)
+                .withCustomPrompt(String.format("Please choose a subscription%s: ",
+                        highlightDefaultValue(defaultValue == null ? null : defaultValue.getSubscriptionName())))
+                .withNumberedPossibleValues(wrapSubs).withDefaultValue(defaultValue).read("Available subscriptions:");
+        if (subscriptionOptionSelected == null) {
+            throw new AzureExecutionException("You must select a subscription.");
+        }
+        return subscriptionOptionSelected.getSubscription();
+    }
+
+    protected Azure createAzureClient() throws AzureAuthFailureException, AzureExecutionException {
         try {
-            final AzureEnvironment environment;
-            if (!AzureAuthHelper.validateEnvironment(auth.getEnvironment())) {
-                Log.prompt(String.format(UNSUPPORTED_AZURE_ENVIRONMENT, auth.getEnvironment()));
-                environment = AzureEnvironment.AZURE;
-            } else {
-                environment = AzureAuthHelper.getAzureEnvironment(auth.getEnvironment());
+            if (this.auth == null) {
+                this.auth = new MavenAuthConfiguration();
+                this.auth.setAuthType(getAuthType());
+            }
+            azureTokenWrapper = MavenAuthHelper.getAzureToken(this.auth, this.session, this.settingsDecrypter).toBlocking().value();
+            if (Objects.isNull(azureTokenWrapper)) {
+                return null;
             }
-            final String environmentName = AzureAuthHelper.getAzureEnvironmentDisplayName(environment);
-            if (environment != AzureEnvironment.AZURE) {
+            final AzureEnvironment env = azureTokenWrapper.getEnv();
+            final String environmentName = AuthHelper.getAzureEnvironmentDisplayName(env);
+            if (env != AzureEnvironment.AZURE) {
                 Log.prompt(String.format(USING_AZURE_ENVIRONMENT, environmentName));
             }
-            azureTokenWrapper = getAuthTypeEnum().getAzureToken(isAuthConfigurationExist() ? this.auth : null, environment);
-            final Azure azureClient = azureTokenWrapper == null ? null : AzureClientFactory.getAzureClient(azureTokenWrapper,
-                    this.subscriptionId, getUserAgent());
-
-            return azureClient;
-        } catch (IOException | AzureLoginFailureException e) {
+            if (StringUtils.isBlank(this.subscriptionId)) {
+                if (StringUtils.isNotBlank(azureTokenWrapper.getDefaultSubscriptionId())) {
+                    this.subscriptionId = azureTokenWrapper.getDefaultSubscriptionId();
+                } else {
+
+                    final TextIO textIO = TextIoFactory.getTextIO();
```

Comment:
seems textIO can be a static field.

Reply:
some time it is not required, so I want to make it lazy loaded.